api/envoy/service/mutation/v3alpha/mutation.proto

```diff
@@ -0,0 +1,275 @@
+
+service Mutation {
+  // This begins the bidirectional stream that Envoy will use to
+  // give the server control over what the filter does. The actual
+  // protocol is described by the MutationRequest and MutationResponse
+  // messages below.
+  rpc Mutate(stream MutationRequest) returns (stream MutationResponse) {
+  }
+}
+
+// This represents the different types of messages that Envoy can send
+// to a mutation server.
+// [#next-free-field: 8]
+message MutationRequest {
+  oneof r {
+    // If true, then Envoy requires that the server send one MutationResponse message
+    // back up the stream in response to this message -- otherwise, Envoy will
+    // never resume processing the filter chain. In practice, this will always
+    // be "true" unless the server or the user has configured the filter
+    // to run in "async mode."
+    bool response_required = 1;
+
+    // Information about the HTTP request, sent when the decodeHeaders callback
+    // is invoked on the Envoy filter.
+    Request request = 2;
+
+    // Information about the HTTP response, sent when the encodeHeaders callback
+    // is invoked on the Envoy filter.
+    Response response = 3;
+
+    // A chunk of the HTTP request body. This is only sent if the server
+    // requested it.
+    Body request_body = 4;
+
+    // A chunk of the HTTP response body, if the server requested it.
+    Body response_body = 5;
+
+    // The HTTP trailers for the request path. These are only sent if
+    // the server requested them.
+    config.core.v3.HeaderMap request_trailers = 6;
+
+    // The HTTP trailers for the response path, sent if the server
+    // requested them.
+    config.core.v3.HeaderMap response_trailers = 7;
+  }
+}
+
+// The mutation server may respond with zero or more of these responses
+// before closing the stream.
+// [#next-free-field: 8]
+message MutationResponse {
+  enum Command {
+    // The server is done with the request and response, and no further
+    // filters should be invoked. This is typically used with the
+    // "CompleteResponse" to return an error.
+    STOP = 0;
+
+    // The server is done with the request. Continue processing
+    // and do not consult the server again for this request.
+    CONTINUE_REQUEST = 1;
+
+    // The server is done with the response. Do not consult the server
+    // again.
+    CONTINUE_RESPONSE = 2;
+
+    // Request that Envoy next send a Body message downstream
+    // containing either the next chunk of the request body available, or an end-of-stream.
+    // Once this is returned, it will no longer be possible to modify the
+    // message headers.
+    SEND_REQUEST_BODY_CHUNK = 3;
+
+    // Send the next chunk of the response body.
+    SEND_RESPONSE_BODY_CHUNK = 4;
+
+    // Request that Envoy will buffer the entire request message body in memory
+    // subject to the currently-configured body buffer limit, and send a
+    // single Body message downstream containing the entire message body.
```

Comment:
What if it runs out of that body buffer limit? The updated Proxy-Wasm ABI will call into request body callbacks again in that case.